README.md
================================================================================
About Gateway_core
================================================================================

The gateway_core is an Exosite one platform gateway application with following features: 

* Automatically activate device client in one platform given proper Vendor/Model/SerialNumber.
* Download user application stored in one platform and launch the application.
* Manage/monitor multiple user applications.
* Restart individual application under user's control.
* Remotely restart the gateway_core.
* Ability to upgrade gateway_core itself.

License is BSD, Copyright 2012, Exosite LLC (see LICENSE file)

================================================================================
QUICK START
================================================================================

1. Create a new clonable device with whatever data sources are appropriate.

2. In the cloned device, also create a data source using "command" as the alias.

3. Create a new Client Model, using the `RID` of the above device.

4. Add some valid serial numbers to the Client Model

5. Edit "options.cfg" to have correct values based on your Client Model.
   Update the other items to suit:

  [setup]  
  vendor_name = VENDOR_NAME  
  model_name = MODEL_NAME  
  interface_type = NETWORK_INTERFACE*  
  uuid_file = UUID_FILE_LOCATION*  
  check_interval = 5  
  onep_server = m2.exosite.com  
  onep_port = 80  
  activate_url = /provision/activate  
  activate_timeout = 600  
  rpc_url = /api:v1/rpc/process  
  http_timeout = 3  
  datastore_interval = 3  
  app_id = gateway_core  
  log_level = info  
  
  --) NETWORK_INTERFACE can be set to 'eth0' or 'uuid'. If eth0 is used,
      add the MAC address of localhost as a serial number and add a device with
      that serial number. No need to change uuid file location.

  --) If 'uuid' is used as network interface, the uuid_file must be specified to
      indicate the file contaning uuid information structure like:

     [UUID]  
     uuid = 01234UUIDVALUE01234

  [network]  
  platform = janus_400ap  
  network_type = NETWORK_TYPE  
  ping_ip = IP_TO_PING  
  ping_interval = PING_INTERVAL  
  ping_retry = RETRY_COUNT  

  The [network] section is configured to monitor/auto-recover network.
  Currently, this is available only on 'janus_400ap' platform. The core
  will ping the IP_TO_PING for every PING_INTERVAL seconds, if it failed
  for RETRY_COUNT times, then core will try to restart the network.
  The NETWORK_TYPE could be 'gsm' or 'cdma864'

6. Modify the gateway_core_options.cfg file to suit:  
    --) Add any apps the system shoud load to the apps_list line, separating
    the apps by a comma.  The section header MUST match the app_id (
    (above) with the suffix '_config'.  If any apps are added or removed
    from the apps_list, the gateway_core will start or stop the added
    or removed apps, respectively:

   [gateway_core_config]  
   apps_list = app1, app2, app3

    --) Set the gateway_core version number as desired.  The gateway_core
    will try to update itself and then reboot the entire system if it
    finds that its version number changed:

  [gateway_core_config]  
  version = 1

    --) If desired, add a config section for each app.  The section header
    must be named the same name as in the apps_list with the suffix of
    "_config".  The gateway_core will restart the app if it finds that
    the "version" parameter changes.

  [app1_config]  
  version = 1.0


7. Upload the gateway_core_options.cfg file as a content file named
   "gateway_core_options" to the Client Model you created

8. The "apps_template" folder contains example of simple app called
   "template.py".  Use this to create your own apps.

9. Upload your apps to the Client Model as content files, naming them
   with the same name as you listed in the options file above

10. Start running gateway_core:

   $python ./boot.pyc

For Your Information:  
* To reset gateway_core, send 'reset' to the 'command' dataport under the
  provisioned device.

* The "apps" folder will be used to store the application script at run time
  and should not be modified by user.

* The provisioned CIK for each individual device is stored in a hidden
  file called ".exosite" in the python folder.  This file can be read
  by the gateway_core app as well as your worker apps.

* To support update gateway_core itself. First tar *.pyc files as
  "gateway_core.tar" and upload it as a content file named "gateway_core".

    tar cvf gateway_core.tar *.pyc

* Each application should implement its own configuration management.
  For example, each application could read from a datasource under
  the device with the same name as the app_name (e.g. "template").
  This data source could store configuration information in json
  format that each app would use to self-modify its behavior based
  on the configuration of the specific device it is running on.  For
  example, the datasource could contain:

  {
    CUSTOMIZED_CONFIG_DATA,
    "monitor":{"name":MONITOR_NAME}
  }


================================================================================
Gateway Application Development Guideline
================================================================================

1) To develop gateway application, the application should be written in python
   and must define a global function named 'main' with an optional argument
   'wdtobject'.

2) Watch Dog Timer:
   To support this feature, you can edit "watch_dog_timer" line in the application
   config section and give it a number in seconds that is the communication timer
   between gateway_core and appication. If 0 is set, this feature is not enabled.
   Also this number should be greater than the time interval in your loop.
   For example.

     def main(wtdobj=None):
     while True:
       [Do something]
       time.sleep(30)

   then, the watch_dog_timer should be set to be greater than 30.

   [template_config]  
   watch_dog_timer = 35

   The following code should be executed before time.sleep() is called.

    wdtobj.sendMessage(os.getpid())

3) Graceful shutdown:
   When you enbale "Watch Dog Timer", this feature is also enabled. The following
   code should be inserted berore starting application loop.

    appproxy.signalShutdownEvent(kill_app)

   The kill_app is a global list which has single element 'False'.

  Full application example code:

    kill_app = [False]
    def main(wdtobj=None):
    global kill_app
    appproxy.signalShutdownEvent(kill_app)
    while False == kill_app[0]:
      print "This is an empty loop"
      wdtobj.sendMessage(os.getpid())
      time.sleep(10)

4) Starting and Restarting Applications:  
   * Write python code and save it as APP_NAME.py.  
   * Upload it as a content file named APP_NAME.  
   * Edit 'apps_list' to append this APP_NAME in 'gateway application option file'
     and add new config section [APP_NAME_config] for APP_NAME.
       
    [gateway_core_config]  
    version = 1  
    apps_list = APP_NAME

    [APP_NAME_config]  
    watch_dog_timer = 15  
    version = 1.0
  
   * upload this config file.  

  For 'gateway application option file', please refer to step 6 in Installation.

  To restart application, modify 'version' as below and upload again.

  [APP_NAME_config]  
  version = 2.0

5) To restart gateway_core, send 'reset' to the 'command' dataport under the
   provisioned device.

================================================================================
End Of File


CHANGES.txt
CHANGE LOG
==========

0.1.0
-----
+ Initial release.

0.1.1
-----
+ Updated Network Monitoring to support interval specified in config file (seconds) between pings

0.1.2
-----
+ Fixed issue with writing the status. Uses Pyonep Datastore write function instead of record. 
+ Allow options.cfg to set Datastore logging



python2.7/onepv1lib/onep.py
#==============================================================================
# onep.py
# Main API library class for Exosite's Data Platform as exposed over HTTP JSON
# RPC
#==============================================================================
##
## Tested with python 2.6
##
## Copyright (c) 2010, Exosite LLC
## All rights reserved.
##

import sys,httplib
from exceptions import *

try:
  if sys.version_info < (2 , 6):
    json_module= 'python-simplejson'
    import simplejson as json
  else:
    json_module= 'python-json'
    import json
except ImportError:
  print "The package '%s' is required." % json_module
  sys.exit(1)

#===============================================================================
class OnepV1():
#===============================================================================
  headers = {'Content-Type': 'application/json; charset=utf-8'}
#-------------------------------------------------------------------------------
  def __init__(self,host='173.255.209.28',port='80',url='/api:v1/rpc/process',httptimeout=3):
    self.host        = host + ':' + port
    self.url         = url
    self.httptimeout = int(httptimeout)
    self._clientid   = None
    self._resourceid = None

#-------------------------------------------------------------------------------
  def __callJsonRPC(self, clientkey, callrequests):
    auth = self.__getAuth(clientkey)
    jsonreq = {"auth":auth,"calls":callrequests}
    if sys.version_info < (2 , 6):
     conn = httplib.HTTPConnection(self.host)
    else:
     conn = httplib.HTTPConnection(self.host, timeout=self.httptimeout)
    param = json.dumps(jsonreq)    
    try:
      conn.request("POST", self.url, param, self.headers)
    except Exception:
      print sys.exc_info()[0]
      raise JsonRPCRequestException("Failed to make http request.")
    try:
      read = conn.getresponse().read()
    except Exception:
      print sys.exc_info()[0]
      raise JsonRPCResponseException("Failed to get response of request.")
    finally:
      conn.close()
    try:
      res = json.loads(read)
    except:
      raise OnePlatformException("Return invalid response value.")
    if isinstance(res,dict) and res.has_key('error'):
      raise OnePlatformException(res['error'])
    if isinstance(res,list) and len(res) > 0:
      if res[0].has_key('status'):
        if 'ok'  == res[0]['status']:
          if res[0].has_key('result'):
            return True,res[0]['result']
          return True,'ok'
        else:
          return False,res[0]['status']
      if res[0].has_key('error'):
        raise OnePlatformException(res[0]['error'])
    raise OneException("Unknown error")

#-------------------------------------------------------------------------------
  def __composeCall(self,method,argu):
    return [{"id":1,"procedure":method,"arguments":argu}]

#-------------------------------------------------------------------------------
  def __getAuth(self,clientkey):
    if None != self._clientid:
      return {"cik":clientkey,"client_id":self._clientid}
    elif None != self._resourceid:
      return {"cik":clientkey,"resource_id":self._resourceid}
    return {"cik":clientkey}

#-------------------------------------------------------------------------------
  def activate(self,clientkey,codetype,code):
    argu = [codetype, code]
    request = self.__composeCall("activate",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def comment(self, clientkey, rid, visibility, comment):
    argu = [rid, visibility, comment]
    request = self.__composeCall("comment",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def connect_as(self, clientid):
    self._clientid = clientid
    self._resourceid = None

#-------------------------------------------------------------------------------
  def connect_owner(self, resourceid):
    self._resourceid = resourceid
    self._clientid = None

#-------------------------------------------------------------------------------
  def create(self, clientkey, type, desc):
    argu = [type,desc]
    request = self.__composeCall("create",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def deactivate(self,clientkey,codetype,code):
    argu = [codetype, code]
    request = self.__composeCall("deactivate",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def drop(self, clientkey, rid):
    argu = [rid]
    request = self.__composeCall("drop",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def flush(self, clientkey, rid):
    argu = [rid]
    request = self.__composeCall("flush",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def info(self, clientkey, rid, options={}):
    argu = [rid,options]
    request = self.__composeCall("info",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def listing(self, clientkey, types, options):
    argu =[types,options]
    request = self.__composeCall("listing",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def lookup(self, clientkey, type, mapping):
    argu = [type,mapping]
    request = self.__composeCall("lookup",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def map(self, clientkey, rid, alias):
    argu = ["alias",rid,alias]
    request = self.__composeCall("map",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def read(self, clientkey, rid, options):
    argu = [rid,options]
    request = self.__composeCall("read",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def record(self, clientkey, rid, entries, options={}):
    argu = [rid, entries,options]
    request = self.__composeCall("record",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def revoke(self,clientkey,codetype,code):
    argu = [codetype, code]
    request = self.__composeCall("revoke",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def share(self, clientkey, rid, options={}):
    argu = [rid, options]
    request = self.__composeCall("share",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def unmap(self, clientkey, alias):
    argu = ["alias",alias]
    request = self.__composeCall("unmap",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def update(self, clientkey, rid, desc={}):
    argu = [rid,desc]
    request = self.__composeCall("update",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def write(self, clientkey, rid, value, options={}):
    argu = [rid,value,options]
    request = self.__composeCall("write",argu)
    return self.__callJsonRPC(clientkey,request)

#-------------------------------------------------------------------------------
  def writegroup(self, clientkey, entries, options={}):
    argu = [entries,options]
    request = self.__composeCall("write",argu)
    return self.__callJsonRPC(clientkey,request)


python2.7/onepv1lib/datastore.py
#==============================================================================
# datastore.py
# Buffering, caching and batch functions for access to the Exosite Data Platform
# using the JSON RPC API over HTTP.
# This layer was written so that a system with many subcriber/provider tasks
# could simulataneously access the platform efficiently.
#==============================================================================
##
## Tested with python 2.6
##
## Copyright (c) 2010, Exosite LLC
## All rights reserved.
##

import threading,time,sys,logging
from onep import OnepV1
from exceptions import *

# setup default configurations
transport_config = {'host':'m2.exosite.com',
                    'port':'80',
                    'url':'/api:v1/rpc/process',
                    'timeout':3}
datastore_config = {'write_buffer_size':1024,
                    'read_cache_size':1024,
                    'read_cache_expire_time':5,
                    'log_level':'debug'}

logdict = {'debug':logging.DEBUG, 'info':logging.INFO, 'warn':logging.WARN, 'error':logging.ERROR}
logger = logging.getLogger("Datastore")
ch = logging.StreamHandler()
formatter = logging.Formatter("[%(levelname)s %(asctime)s %(funcName)s:%(lineno)d] %(message)s")
ch.setFormatter(formatter)
logger.addHandler(ch)

lock = threading.Lock()

#==============================================================================
class Datastore():
#==============================================================================
#-------------------------------------------------------------------------------
  def __init__(self,cik,interval,autocreate=False,config=datastore_config,transport=transport_config):
    self._liveBuffer = dict()
    self._recordBuffer = dict()
    self._aliasDict = dict()
    self._cache = dict()
    self._cacheCount = 0
    self._recordCount = 0
    self._auto = autocreate
    self._config = config
    self._conn = OnepV1(transport['host'],transport['port'],transport['url'],transport['timeout'])
    self._cik  = cik
    if interval < 1:
      interval = 1
    self._interval = interval
    logger.setLevel(logdict[self._config['log_level']])

#-------------------------------------------------------------------------------
  def __bufferCount(self):
    return len(self._liveBuffer) + self._recordCount

#-------------------------------------------------------------------------------
  def __isBufferFull(self):
    return self.__bufferCount() >= self._config['write_buffer_size']

#-------------------------------------------------------------------------------
  def __isLiveBufferEmpty(self):
    if self._liveBuffer:
      return False
    else:
      return True

#-------------------------------------------------------------------------------
  def __forceTerminate(self):
    if self._killed and self._forceterminate:
      self._liveBuffer.clear()
      return True
    else:
      return False

#==============================================================================
# One platform queries below
#==============================================================================
#-------------------------------------------------------------------------------
  def __lookup(self,alias,forcequery=False):
    if (not forcequery) and self._aliasDict.has_key(alias):
      return self._aliasDict[alias]
    else:
      status,res = self._conn.lookup(self._cik,'alias',alias)
      if not status:
        self._aliasDict[alias] = False
        return False
      else:
        self._aliasDict[alias] = res
        return res

#-------------------------------------------------------------------------------
  def __read(self,alias,count=1,forcequery=False,sort='desc',starttime=None,endtime=None):
    rid = self.__lookup(alias,forcequery)
    if None != starttime and None != endtime:
      status,res = self._conn.read(self._cik,rid,{"starttime":starttime,"endtime":endtime,"limit":count,"sort":sort})
    else:
      status,res = self._conn.read(self._cik,rid,{"limit":count,"sort":sort})
    if not status:
      raise OneException("Error message from one platform (read): %s" % res)
    return res

#-------------------------------------------------------------------------------
  def __record(self,alias,entries):
    rid = self.__lookup(alias)
    record_status,record_message = self._conn.record(self._cik,rid,entries)
    if not ( True == record_status and 'ok' == record_message):
      raise OneException("Error message from one platform (record): %s" % record_message)
    return True

#-------------------------------------------------------------------------------
  def __writegroup(self,entries):
    data = list()
    for (alias,value) in entries:
      rid = self.__lookup(alias)
      data.append([rid,value])
    write_status,write_message = self._conn.writegroup(self._cik,data)
    if not (True == write_status and 'ok' == write_message):
      raise OneException("Error message from one platform (write): %s,%s" % (value,write_message))
    return True

#-------------------------------------------------------------------------------
  def __createDataport(self,alias,name=None,format="string",preprocess=[],count="infinity",duration="infinity",visibility='parent'):
    if None == name:
      name = alias
    desc = {"format":format,"name":name,'visibility':visibility,"retention":{"count":count,"duration":duration},"preprocess":preprocess}
    create_status,rid = self._conn.create(self._cik,"dataport",desc)
    if create_status:
      map_status,map_message = self._conn.map(self._cik,rid,alias)
      if map_status:
        self._aliasDict[alias] = rid
        return True
      else:
        self._conn.drop(self._cik,rid)
        logger.error(map_message)
        return False
    else:
      logger.error(rid)
      return False

#-------------------------------------------------------------------------------
  def __checkDataportExist(self,alias):
    if self.__lookup(alias):
      return True
    else:
      if self._auto:
        if not self.__createDataport(alias=alias,format=self._auto['format'],preprocess=self._auto['preprocess'],count=self._auto['count'],duration=self._auto['duration'],visibility=self._auto['visibility']):
          raise OneException("Fail to create dataport.")
        return True
      else:
        logger.warn("Data source does not exist while not in AUTO_CREATE mode.")
        return False

#==============================================================================
# Write buffer processing below
#==============================================================================
#-------------------------------------------------------------------------------
  def __processJsonRPC(self):
    while not self.__forceTerminate():
      time.sleep(self._interval)
      livedata = list()
      lock.acquire()
      try:
        timestamp = int(time.time())
        aliases = self._liveBuffer.keys()
        for alias in aliases:
          value = self._liveBuffer[alias]
          try:
            if self.__checkDataportExist(alias): # create datasource if necessary
              livedata.append([alias,value]) # Move to live data
              logger.debug("Data to be written (alias,value): ('%s',%s)" % (alias,value) )
          except OneException: #catch exception, add to recordBuffer
            if not self._recordBuffer.has_key(alias):
                self._recordBuffer[alias] = list()
            self._recordBuffer[alias].append([timestamp,value,True])
            self._recordCount += 1
          finally:
            del self._liveBuffer[alias]
      finally:
        self._liveBuffer.clear()
        lock.release()
      # write live data
      if livedata:
        timestamp = int(time.time())
        try:
          self.__writegroup(livedata)
          logger.info("[Live] Written to 1p:" + str(livedata))
        except OneException,e: # go to historical data when write live data failure
          logger.warn(e.message)
          lock.acquire()
          try:
            for (alias,value) in livedata:
              if not self._recordBuffer.has_key(alias):
                self._recordBuffer[alias] = list()
              offset = True
              self._recordBuffer[alias].append([timestamp,value,offset])
              self._recordCount += 1
          finally:
            lock.release()
        except Exception,e:
          print sys.exc_info()[0]
      ## write historical data
      lock.acquire()
      try:
        aliases =  self._recordBuffer.keys()
        curtime = int(time.time())
        for alias in aliases:
          entries = self._recordBuffer[alias]
          try:
            if self.__checkDataportExist(alias):
              recentry = list()
              for entry in entries:
                if True == entry[2]: #offset mode
                  offset = entry[0] - curtime
                  recentry.append([offset, entry[1]])
                else:
                  recentry.append([entry[0], entry[1]])
              if recentry:
                try:
                  self.__record(alias,recentry)
                  logger.info("[Historical] Written to 1p: " + alias + ", " + str(recentry))
                  self._recordCount -= len(entries)
                  del self._recordBuffer[alias]
                except OneException,e:                  
                  if isinstance(e.message, dict):                    
                    continue
                  if e.message.find("datapoint") != -1:                    
                    self._recordCount -= len(entries)
                    del self._recordBuffer[alias]
            else:
              del self._recordBuffer[alias]
          except OneException,e:
            logger.warn(e.message)
            continue
      finally:
        lock.release()
      if self._killed and not self._recordBuffer:
        self._forceterminate = True

#==============================================================================
# Read cache routines below
#==============================================================================
#-------------------------------------------------------------------------------
  def __addCacheData(self,alias,count,forcequery=False):
    if self.__isCacheFull():
      self.__clearCache()
    self._cache[alias] = dict()
    data = self.__refreshData(alias,count,forcequery)
    if data:
      self._cacheCount += 1
    return data

#-------------------------------------------------------------------------------
  def __isExpired(self,alias):
    try:
      return int(time.time()) - self._cache[alias]['time'] > self._config['read_cache_expire_time']
    except:
      return True

#-------------------------------------------------------------------------------
  def __isCacheFull(self):
    return self._cacheCount >= self._config['read_cache_size']

#-------------------------------------------------------------------------------
  def __clearCache(self):
    self._cache.clear()
    self._cache = dict()
    self._cacheCount = 0

#-------------------------------------------------------------------------------
  def __refreshData(self,alias,count,forcequery=False):
    try:
      time.sleep(1)
      data = self.__read(alias,count,forcequery)
      self._cache[alias]['data'] = data
      self._cache[alias]['time'] = int(time.time())
      return data
    except OneException,e:
      logger.warn(e.message)
    except Exception,e:
      print sys.exc_info()[0]
      logger.error("Unknown error when reading data.")
    return False

#==============================================================================
# Public methods below
#==============================================================================
#-------------------------------------------------------------------------------
  def isThreadAlive(self):
    return self._thread.isAlive()

#-------------------------------------------------------------------------------
  def comment(self,alias,visibility,commentstr):
    rid = self.__lookup(alias)
    if rid:
      status,message = self._conn.comment(self._cik,rid,visibility,commentstr)
      if status:
        return True
    return False

#-------------------------------------------------------------------------------
  def createDataport(self,alias,format,name=None,preprocess=[],count=0,duration=0,visibility="public"):
    rid = self.__lookup(alias)
    if rid:
      return False,"Alias already existed."
    else:
      if self.__createDataport(alias,name,format,preprocess,count,duration,visibility):
        return True,True
      else:
        return False,"Failed to create Dataport."

#-------------------------------------------------------------------------------
  def read(self,alias,count=1,forcequery=False):
    if self._cache.has_key(alias): #has cache data
      if self.__isExpired(alias) or count != len(self._cache[alias]['data']):
        return self.__refreshData(alias,count,forcequery)
      else:
        return self._cache[alias]['data']
    else: #no cache data
      return self.__addCacheData(alias,count,forcequery)

#-------------------------------------------------------------------------------
  def record(self,alias,entries):
    if self.__isBufferFull() or not (self._auto or self.__lookup(alias)):
      return False
    lock.acquire()
    try:
      if not self._recordBuffer.has_key(alias):
        self._recordBuffer[alias] = list()
      for (t,value) in entries:
        recentry = [t,value,False]
        self._recordBuffer[alias].append(recentry)
    finally:
      lock.release()

#-------------------------------------------------------------------------------
  def restart(self):
    self.stop(force=True)
    self.start()

#-------------------------------------------------------------------------------
  def start(self,daemon=False):
    time.sleep(1)
    self._killed = False
    self._forceterminate = False
    self._thread = threading.Thread(target=self.__processJsonRPC)
    self._thread.setDaemon(daemon)
    self._thread.start()

#-------------------------------------------------------------------------------
  def stop(self,force=False):
    self._killed = True
    self._forceterminate = force

#-------------------------------------------------------------------------------
  def write(self,alias,value):
    if self.__isBufferFull() or not (self._auto or self.__lookup(alias)):
      return False
    else:
      lock.acquire()
      try:
        if self._liveBuffer.has_key(alias):
          self._liveBuffer[alias] = value
          logger.debug("Update the (alias,value) in buffer:%s,%s" % (alias,value))
          return False
        else:
          self._liveBuffer[alias] = value
      finally:
        lock.release()
      logger.debug("Current buffer count: %s" % self.__bufferCount())
      logger.debug("Add to buffer:%s,%s" % (alias,value))
      return True


python2.7/apps_template/template.py
#==============================================================================
# template.py
# The is a template to start coding other apps that are part of the
# gateway core application framework.
#==============================================================================
##
## Tested with python 2.6
##
## Copyright (c) 2011, Exosite LLC
## All rights reserved.
##
import sys, time, os
import appproxy
from onepv1lib.datastore import Datastore
from onepv1lib.provision import Provision

kill_app = [False]

#===============================================================================
def main(wdtobj=None):
#===============================================================================
  global kill_app
  utility = UtilityFunctions()
  #print "Reading from Exosite:",utility.readData("status")
  #print utility.getModelContent('template')
  ### Uncomment following line if 'watch_dog_timer' in app option file is setup to > 0.
  appproxy.signalShutdownEvent(kill_app)
  while False == kill_app[0]:
    try:
      print "This is an empty loop. pid:", os.getpid()
      #utility.writeData("test", 6)
      ### Uncomment following line if 'watch_dog_timer' in app option file is setup to > 0.
      wdtobj.sendMessage(os.getpid())
      time.sleep(10)
    except (KeyboardInterrupt, SystemExit):
      kill_app[0] = True
  print "Leaving application"

#===============================================================================
def stop():
#===============================================================================
  global kill_app
  kill_app = True

#===============================================================================
class UtilityFunctions():
#===============================================================================
  def __init__(self):
    self._conparams = {'onep_server':'m2.exosite.com','onep_port':"80",'rpc_url':'/api:v1/rpc/process', 'http_timeout':3, 'datastore_interval':3}
    self._modelparams = {'vendor_name':'intime','model_name':'intime_gateway'}
    self._cik = self._readCIK()
    self._datastore = None
    self._initDataStore()
    self._prov = Provision('http://' + self._conparams["onep_server"]
                           + ':' + self._conparams["onep_port"])

#========================== public methods =====================================
#-------------------------------------------------------------------------------
  def getCIK(self):
    return self._cik

  def writeData(self, alias, value):
    try:
      return self._datastore.write(alias, value)
    except:
      pass
#-------------------------------------------------------------------------------
  def readData(self, alias):
    try:
      return self._datastore.read(alias)
    except:
      pass

#-------------------------------------------------------------------------------
  def getModelContent(self,contentname):
    return self._getContent(contentname)

#========================== private methods ====================================
#------------------------------------------------------------------------------
  def _readCIK(self):
    path = ".exosite"
    if os.path.isfile(path):
      try:
        file = open(path)
      except:
        print "CIK file not found"
        return None
      cik = file.read()
      file.close()
      if len(cik) == 40:
        return cik
      else:
        print "Invalid CIK (length)"
        return None
    else:
      print "CIK file not found"
      return None

#------------------------------------------------------------------------------
  def _initDataStore(self):
    transport_config = {
                        'host':self._conparams['onep_server'],
                        'port':self._conparams['onep_port'],
                        'url': self._conparams['rpc_url'],
                        'timeout':self._conparams['http_timeout']
                       }
    if None != self._cik:
      self._datastore = Datastore(
                                  self._cik,
                                  self._conparams['datastore_interval'],
                                  autocreate={'format':'integer', 'preprocess':[], 'count':'infinity', 'duration':'infinity', 'visibility':'parent'},
                                  transport=transport_config
                                 )
      self._datastore.start()

#------------------------------------------------------------------------------
  def _getContent(self,contentid):
    return self._prov.content_download(
                                       self.getCIK(),
                                       self._modelparams["vendor_name"],
                                       self._modelparams["model_name"],
                                       contentid
                                      )

#===============================================================================
if __name__ == '__main__':
  try:
    main()
  except (KeyboardInterrupt, SystemExit):
    sys.exit(0)
  finally:
    stop()


